File: javascript/basics/01_syntax.js
/**
 * JavaScript Basics - Cú pháp cơ bản
 */

// 1. Comments (Ghi chú)
// Single line comment
/* Multi-line comment */

// 2. Console output
console.log("Hello, World!");
console.log("Xin chào", "JavaScript");

// 3. Variables
// var (old way, function-scoped)
var oldVar = "old";

// let (block-scoped, can be reassigned)
let name = "JavaScript";
name = "JS";

// const (block-scoped, cannot be reassigned)
const PI = 3.14159;
// PI = 3.14; // Error!

// 4. Data Types
// String
let text = "Hello";
let text2 = 'World';
let template = `Template literal ${text}`;

// Number
let integer = 42;
let float = 3.14;
let infinity = Infinity;
let notANumber = NaN;

// Boolean
let isTrue = true;
let isFalse = false;

// Null and Undefined
let nothing = null;
let notDefined = undefined;

// Symbol (ES6)
let sym = Symbol("unique");

// BigInt (ES2020)
let bigNumber = 9007199254740991n;

// 5. Type checking
console.log(typeof text);        // "string"
console.log(typeof integer);     // "number"
console.log(typeof isTrue);      // "boolean"
console.log(typeof nothing);     // "object" (bug in JS)
console.log(typeof notDefined);  // "undefined"

// 6. Type conversion
let strNumber = "123";
let intNumber = parseInt(strNumber);
let floatNumber = parseFloat("3.14");
let numberFromString = Number("42");
let stringFromNumber = String(42);
let boolFromString = Boolean("true");

// 7. Template literals (ES6)
let firstName = "John";
let lastName = "Doe";
let fullName = `${firstName} ${lastName}`;
console.log(fullName); // "John Doe"

// 8. Operators
// Arithmetic
let sum = 5 + 3;
let difference = 5 - 3;
let product = 5 * 3;
let quotient = 5 / 3;
let remainder = 5 % 3;
let power = 5 ** 3; // ES2016

// Comparison
let isEqual = 5 === 5;      // Strict equality
let isNotEqual = 5 !== 3;   // Strict inequality
let isLooseEqual = 5 == "5"; // Loose equality (avoid!)

// Logical
let and = true && false;
let or = true || false;
let not = !true;

// 9. Nullish coalescing (ES2020)
let value = null ?? "default";
let value2 = undefined ?? "default";

// 10. Optional chaining (ES2020)
let obj = { name: "John" };
let name = obj?.name;
let age = obj?.age?.value;

File: python/basics/01_syntax.py
"""
Python Basics - Cú pháp cơ bản
"""

# 1. Comments (Ghi chú)
# Single line comment
"""
Multi-line comment
hoặc docstring
"""

# 2. Print statement
print("Hello, World!")
print("Xin chào", "Python")

# 3. Variables (Biến)
name = "Python"
age = 30
height = 1.75
is_student = True

# 4. Data Types (Kiểu dữ liệu)
# String
text = "Hello"
# Integer
number = 42
# Float
decimal = 3.14
# Boolean
is_true = True
is_false = False
# None
nothing = None

# 5. Type checking
print(type(name))      # <class 'str'>
print(type(age))       # <class 'int'>
print(type(height))    # <class 'float'>
print(type(is_student)) # <class 'bool'>

# 6. Type conversion
str_number = "123"
int_number = int(str_number)
float_number = float(str_number)

# 7. String formatting
name = "Python"
print(f"Hello, {name}!")  # f-string (Python 3.6+)
print("Hello, {}!".format(name))  # .format()
print("Hello, %s!" % name)  # % formatting

# 8. Input from user
# user_input = input("Nhập tên của bạn: ")
# print(f"Xin chào, {user_input}!")

